rfc/0084-app-service-level-telemetry.md

```diff
@@ -0,0 +1,517 @@
+specific and these may not have a similar shape as the services were
+independently developed (among other reasons).  There are a number of other
+errors, like InvalidArgument and KeyNotFound which are not useful from a
+Capella Experience perspective.
+
+The three metrics we should collect are:
+
+* Ambiguous Timeout
+  * The requested operation has timed out, but the SDK does not know if it was
+  received by the intended remote system.  For example, it was written to the
+  network and timed out before the response was received.  This is only for
+  mutations, as get() operations will always return an UnambiguiousTimeout.
+* Unambiguous Timeout
+  * The requested operation has timed out and the SDK knows it could not have
+  been received by the intended remote system.  For example, it was intended
+  for a system which there was no connection for and the connection couldn't be
+  built to send it before the connection timeout.
+* Request Canceled
+  * The requested operation was started and dispatched to the intended system.
+  However, it has been canceled by the SDK's processing.  Typically this is if
+  the connection is severed, but it could also have other causes (e.g.,
+  shutdown) and can even be triggered by the user themselves (e.g., calling
+  `.cancel()` on a result from another thread)
+
+An additional metric for total requests completed would allow to account for
+other types of errors.
+
+For this set of metrics, we propose gathering simple counters, and on a per SDK
+instance basis (these do not correlate 1:1 to hostnames and a definition of
+runtime/instance identifiers is in the [Response Time Observability
+RFC](rfc/0035-rto.md)). Each counter must be associated with the connection and
+bucket if it is possible. In particular, it must be possible to label counters
+with bucket name and hostname for KV service, while often not possible to even
+reliably associate hostname with Search request due to proxies on the Cloud
+deployments.
+
+From these two simple pieces of information, we can intuit that the connection
+was likely severed and a number of operations timed out before the connection
+could be re-established.
+
+Determining why the connection was severed requires correlation to other
+information.  For example, we may need to look at the Couchbase Server logs to
+see that a particular process for a given service crashed where we would expect
+to find the SDK runtime/instance identifier being sent in the request (defined
+in the [Response Time Observability RFC](rfc/0035-rto.md)).
+
+We should note that existence of timeouts is possibly expected.  For example,
+with an ad-targeting or recommendation targeting deployment, there may be short
+latencies that when not serviced, lead to other processing.  This is worth
+mentioning because it will be important to compare a deployment to itself, and
+there may be limited value in comparing across deployments.
+
```

Comment:
For the Capella Service Level Metrics we only record a subset of SDK operations (the ones listed in the section below) – Eventing management operations are excluded for example. Do we want to count _all_ operations, including all management operations, for the Capella Service Experience Counters?